public/app/features/panel/metrics_panel_ctrl.ts

```diff
@@ -79,9 +79,28 @@ class MetricsPanelCtrl extends PanelCtrl {
     delete this.error;
     this.loading = true;
 
+    // set "mydatasource" to whatever the panel has defined
+    let mydatasource = this.panel.datasource;
+    let datasourceVarName = '';
+
+    // look for data source variables
+    for (let i = 0; i < this.templateSrv.variables.length; i++) {
+      const variable = this.templateSrv.variables[i];
+      if (variable.type !== 'datasource') {
+        continue;
+      }
+
+      datasourceVarName = variable.name;
+    }
+
+    // if a data source variable was found, use its value
+    if (datasourceVarName !== '' && this.panel.scopedVars && this.panel.scopedVars[datasourceVarName]) {
+      mydatasource = this.panel.scopedVars[datasourceVarName].value;
+    }
+
     // load datasource service
     this.datasourceSrv
-      .get(this.panel.datasource)
+      .get(mydatasource)
```

Comment:
would it not be easier to skip all the added code above and pass scopedVars to datasourceSrv.get ?

Reply:
I just decided to learn enough ts to hack this change out as we desperately need to stop littering our dashboards with duplicated, "hardcoded" datasources.  I agree that moving the code to datasource_srv would be preferable, but this isn't the only place datasourceSrv.get() is called, and typescript's method overloading options seem a bit weak.

Reply:
> and typescript's method overloading options seem a bit weak. 

??? 

Well it has do be done. It cannot be done here.

Reply:
think you did not understand what I meant, you can see my refactoring & simplification here: https://github.com/grafana/grafana/pull/15914/files